airflow/models/baseoperator.py

```diff
@@ -365,6 +373,62 @@ def partial(
     )
 
 
+class ExecutorSafeguard:
+    """
+    The ExecutorSafeguard decorator.
+
+    Checks if the execute method of an operator isn't manually called outside
+    the TaskInstance as we want to avoid bad mixing between decorated and
+    classic operators.
+    """
+
+    test_mode = conf.getboolean("core", "unit_test_mode")
+
+    @classmethod
+    def decorator(cls, func):
+        def find_task_instance(frame: FrameSummary):
+            return (
+                frame.name == _execute_task.__name__
+                and _execute_task.__module__.replace(".", os.path.sep) in frame.filename
+            )
+
+        def find_operator(frame: FrameSummary):
+            from airflow.decorators.base import DecoratedOperator
+            from airflow.operators.python import PythonOperator
+
+            return (
+                frame.name == PythonOperator.execute.__name__
+                and PythonOperator.__module__.replace(".", os.path.sep) in frame.filename
+            ) or (
+                frame.name == DecoratedOperator.execute.__name__
+                and DecoratedOperator.__module__.replace(".", os.path.sep) in frame.filename
+            )
+
+        def raise_or_warn(operator: BaseOperator, message: str):
+            if not operator.allow_mixin:
+                raise AirflowException(message)
+            operator.log.warning(message)
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            if not cls.test_mode:
```

Comment:
We should probably look at `self.test_mode` too https://github.com/apache/airflow/blob/83f8a2ec6fc0643bf31d165c013ddf72a120220d/airflow/models/taskinstance.py#L2411

Reply:
I've tried this but a lot of unit tests start failing when doing the above suggestion, so I reverted that change